Fix emitting events for pods preempted by Kueue [pod groups]

mimowo opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

/kind cleanup
/kind bug
<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:

Solve two issues with events for pod groups:
1. the event indicating the a pod in pod group was emitted repeatedly, while the intention is to sent it once
2. the event was emitted only for the first pod in group, but because the Load function may return the pods in different order, we could get both pods covered with at least one event

This PR also adds e2e test to ensure the events are emitted per pod in the group.

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".
-->
Fix for unreleased feature
```release-note
NONE
```

mimowo commented:
/cc @alculquicondor

mimowo commented:
/test pull-kueue-test-e2e-main-1-27
MultiKueue failure

mimowo commented:
/assign @yaroslava-serdiuk

alculquicondor commented:
/retest

alculquicondor commented on test/e2e/singlecluster/pod_test.go:
```diff
@@ -273,6 +275,16 @@ var _ = ginkgo.Describe("Pod groups", func() {
 
 		ginkgo.It("should allow to preempt the lower priority group", func() {
+			podList := corev1.PodList{}
+			podWatcher, err := k8sClient.Watch(ctx, &podList, &client.ListOptions{
```
The watcher should be for Events, not for Pods.

mimowo replied:
I see, let me try that.

mimowo replied:
Done, PTAL

alculquicondor commented:
Please check the failure

mimowo commented:
for the record: https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_kueue/1667/pull-kueue-test-e2e-main-1-29/1752753221249536000

mimowo commented:
I reproduce this failure locally. What is more it seems these events are really not always sent. Locally I run the `kubectl get events -npod-e2e` (fixed the namespace for ease) and these are the events:

```
DELETED    37s         Warning   ErrWorkloadCompose   pod/default-priority-group-0                   'default-priority-group' group total count is less than the actual number of pods in the cluster
DELETED    37s         Normal    CreatedWorkload      pod/default-priority-group-0                   Created Workload: pod-e2e/default-priority-group
DELETED    37s         Normal    Started              pod/default-priority-group-0                   Admitted by clusterQueue cq
DELETED    37s         Normal    Scheduled            pod/default-priority-group-0                   Successfully assigned pod-e2e/default-priority-group-0 to kind-worker
DELETED    35s         Normal    Stopped              pod/default-priority-group-0                   Preempted to accommodate a higher priority Workload
DELETED    37s         Normal    Pulled               pod/default-priority-group-0                   Container image "gcr.io/k8s-staging-perf-tests/sleep:v0.1.0" already present on machine
DELETED    37s         Normal    Created              pod/default-priority-group-0                   Created container c
DELETED    36s         Normal    Started              pod/default-priority-group-0                   Started container c
DELETED    37s         Normal    Started              pod/default-priority-group-1                   Admitted by clusterQueue cq
DELETED    37s         Normal    Scheduled            pod/default-priority-group-1                   Successfully assigned pod-e2e/default-priority-group-1 to kind-worker
DELETED    37s         Normal    Pulled               pod/default-priority-group-1                   Container image "gcr.io/k8s-staging-perf-tests/sleep:v0.1.0" already present on machine
DELETED    37s         Normal    Created              pod/default-priority-group-1                   Created container c
DELETED    36s         Normal    Started              pod/default-priority-group-1                   Started container c
```
Note that there is no `Preempted to accommodate a higher priority Workload` for `pod/default-priority-group-1`. 

Also, I checked with another command that both pods are actually preempted (I get the `TerminationTarget` for both of them).

I will investigate tomorrow why is this so. In the meanwhile, let me know @alculquicondor if you have some insights.

alculquicondor commented:
I would first question whether Kueue is somehow skipping recording an event for certain pods.... or we are just facing the nature of events: writing events is not something that the client retries.

If the former, then we can fix it. If the later, I guess it's worse than I thought and we can only write reliable unit test, but not integration tests. Although I seem to remember we have reliable integration tests for events in kube-scheduler (under the assumption that event recording failures are highly uncommon).

Can you try to verify that?

alculquicondor commented:
Yup, we have reliable integration tests for events in kube-scheduler https://github.com/kubernetes/kubernetes/blob/57e9981dc2745b2f99d01bf0df6bedb0f16ed051/test/integration/scheduler/scheduler_test.go#L286

So we need to check the reconciler code that somehow isn't skipping events for particular pods.

alculquicondor commented:
We can probably accept a certain level of flakiness for this test, but it should be less than 1%

alculquicondor commented on test/e2e/singlecluster/pod_test.go:
```diff
@@ -326,6 +341,24 @@ var _ = ginkgo.Describe("Pod groups", func() {
+						event, ok := evt.Object.(*v1.Event)
+						gomega.Expect(ok).To(gomega.BeTrue())
+						objKey := types.NamespacedName{Namespace: event.InvolvedObject.Namespace, Name: event.InvolvedObject.Name}
+						if defaultGroupPods.Has(objKey) && event.Reason == "Stopped" && event.Message == "Preempted to accommodate a higher priority Workload" {
```
let's not compare the message.

mimowo replied:
done

alculquicondor reviewed (commented):
The test itself looks about right
/approve

mimowo commented:
> I would first question whether Kueue is somehow skipping recording an event for certain pods.... or we are just facing the nature of events: writing events is not something that the client retries.

Kueue is skipping sending of the events. I added temporarily log line around https://github.com/kubernetes-sigs/kueue/blob/a8c11f351858cd159a3f823b15860918c0490cbc/pkg/controller/jobframework/reconciler.go#L748, and can confirm this is ocasionally not called for ` pod/default-priority-group-1`. It is always called for ` pod/default-priority-group-0`. Looking for the solution.

mimowo commented:
Ok, I understand the root cause. We assume with the stop interface that the only stopped Job is represented with the object, but this is not the case for composable jobs, where multiple pods are stopped.

Also, I know why it sometimes passes, this is because the `stopJob` routine is called multiple times before the workload finished (around 10 times), and the `Load` function returns the pods in a non-deterministic order. So sometimes we get the event for `pod-1`.

There is another issue, we actually send the `Stopped` now event multiple times for the same pod, because the Stop function returns always true, assuming no errors: https://github.com/kubernetes-sigs/kueue/blob/a8c11f351858cd159a3f823b15860918c0490cbc/pkg/controller/jobs/pod/pod_controller.go#L462.

I think one solution would be to extend the API of Stop by returning a slice of actually stopped Jobs "now", or slice of extra jobs stopped now. WDYT?

mimowo commented:
> We can probably accept a certain level of flakiness for this test, but it should be less than 1%

After the fix it is quite stable, looped it locally and passed >200 times in a row .

yaroslava-serdiuk commented on pkg/controller/jobs/job/job_controller.go:
```diff
@@ -157,14 +157,14 @@ func (j *Job) Suspend() {
+			return stoppedNow, fmt.Errorf("suspend: %w", err)
 		}
-		stoppedNow = true
+		stoppedNow = []client.Object{j.Object()}
```
Why do we need slice if we return 0 or 1 object?

mimowo replied:
In the current approach I implement the `Stop` function from the `JobWithCustomStop` interface. 

This interface is also used by pod groups, where we want to emit the events about multiple objects (pods). 

I guess we could have the `Stop` function under `ComposableJob`. WDYT? cc @alculquicondor

mimowo replied:
I just pushed a commit which reverts the change for the Job controller, and introduces a Stop function in the ComposableJob. I think it looks better, PTAL.

alculquicondor commented on pkg/controller/jobs/pod/pod_controller_test.go:
```diff
@@ -2451,6 +2764,9 @@ func TestReconciler(t *testing.T) {
 			if diff := cmp.Diff(tc.wantWorkloads, gotWorkloads.Items, tc.workloadCmpOpts...); diff != "" {
 				t.Errorf("Workloads after reconcile (-want,+got):\n%s", diff)
 			}
+			if diff := cmp.Diff(tc.wantEvents, recorder.RecordedEvents, cmp.Transformer("sortedEvents", utiltesting.SortedEvents)); diff != "" {
```
There is a specific opt for sorting :) https://pkg.go.dev/github.com/google/go-cmp/cmp/cmpopts#SortSlices

mimowo replied:
Indeed, opened a small cleanup https://github.com/kubernetes-sigs/kueue/pull/1691

alculquicondor reviewed (commented):
/lgtm
/approve
